File: apps/receiver/native-modules/android/BLEBeaconScanner.kt
package com.phoenix.receiver

import android.bluetooth.BluetoothAdapter
import android.bluetooth.BluetoothManager
import android.bluetooth.le.BluetoothLeScanner
import android.bluetooth.le.ScanCallback
import android.bluetooth.le.ScanResult
import android.bluetooth.le.ScanSettings
import android.content.Context
import android.util.Log
import com.facebook.react.bridge.*
import com.facebook.react.modules.core.DeviceEventManagerModule

class BLEBeaconScanner(reactContext: ReactApplicationContext) :
    ReactContextBaseJavaModule(reactContext) {

    private var bluetoothAdapter: BluetoothAdapter? = null
    private var bluetoothLeScanner: BluetoothLeScanner? = null
    private var isScanning = false
    private val reactApplicationContext: ReactApplicationContext = reactContext

    companion object {
        private const val TAG = "BLEBeaconScanner"
        private const val PHOENIX_COMPANY_ID_SAMSUNG = 0x0075 // Samsung (Android emitter)
        private const val PHOENIX_COMPANY_ID_APPLE = 0x004C // Apple (iOS emitter)
    }

    override fun getName(): String {
        return "BLEBeaconScanner"
    }

    init {
        val bluetoothManager =
            reactContext.getSystemService(Context.BLUETOOTH_SERVICE) as BluetoothManager
        bluetoothAdapter = bluetoothManager.adapter
        bluetoothLeScanner = bluetoothAdapter?.bluetoothLeScanner
    }

    private val scanCallback = object : ScanCallback() {
        override fun onScanResult(callbackType: Int, result: ScanResult) {
            super.onScanResult(callbackType, result)
            processScanResult(result)
        }

        override fun onBatchScanResults(results: MutableList<ScanResult>) {
            super.onBatchScanResults(results)
            for (result in results) {
                processScanResult(result)
            }
        }

        override fun onScanFailed(errorCode: Int) {
            super.onScanFailed(errorCode)
            isScanning = false
            val errorMsg = "BLE scan failed: ${getScanErrorMessage(errorCode)}"
            Log.e(TAG, errorMsg)
            NativeLogger.error(TAG, errorMsg)

            sendEvent("onScanningStateChange", Arguments.createMap().apply {
                putBoolean("scanning", false)
                putString("error", getScanErrorMessage(errorCode))
            })
        }
    }

    @ReactMethod
    fun initializeScanner(promise: Promise) {
        try {
            if (bluetoothAdapter == null) {
                promise.reject("BLE_NOT_SUPPORTED", "Bluetooth is not supported on this device")
                return
            }

            if (!bluetoothAdapter!!.isEnabled) {
                promise.reject("BLE_DISABLED", "Bluetooth is disabled")
                return
            }

            if (bluetoothLeScanner == null) {
                promise.reject("BLE_SCANNER_UNAVAILABLE", "BLE scanner is not available")
                return
            }

            val result = Arguments.createMap().apply {
                putString("state", "poweredOn")
                putBoolean("initialized", true)
            }

            promise.resolve(result)
        } catch (e: Exception) {
            promise.reject("INIT_ERROR", "Failed to initialize scanner: ${e.message}")
        }
    }

    @ReactMethod
    fun startScanning(promise: Promise) {
        try {
            if (isScanning) {
                promise.resolve(Arguments.createMap().apply {
                    putBoolean("scanning", true)
                })
                return
            }

            if (bluetoothLeScanner == null) {
                promise.reject("BLE_UNAVAILABLE", "BLE scanner is not available")
                return
            }

            // Configure scan settings for optimal performance
            val settings = ScanSettings.Builder()
                .setScanMode(ScanSettings.SCAN_MODE_LOW_LATENCY) // Most responsive
                .setReportDelay(0) // Report immediately, no batching
                .setCallbackType(ScanSettings.CALLBACK_TYPE_ALL_MATCHES) // Report all matches
                .setMatchMode(ScanSettings.MATCH_MODE_AGGRESSIVE) // More aggressive matching
                .setNumOfMatches(ScanSettings.MATCH_NUM_MAX_ADVERTISEMENT) // Max advertisements
                .build()

            // Start scanning (no filters - we'll filter in callback)
            bluetoothLeScanner!!.startScan(null, settings, scanCallback)
            isScanning = true

            Log.d(TAG, "BLE beacon scanning started")
            NativeLogger.info(TAG, "BLE beacon scanning started")

            sendEvent("onScanningStateChange", Arguments.createMap().apply {
                putBoolean("scanning", true)
            })

            promise.resolve(Arguments.createMap().apply {
                putBoolean("scanning", true)
            })

        } catch (e: SecurityException) {
            promise.reject("PERMISSION_DENIED", "Bluetooth permissions not granted: ${e.message}")
        } catch (e: Exception) {
            promise.reject("SCAN_ERROR", "Failed to start scanning: ${e.message}")
        }
    }

    @ReactMethod
    fun stopScanning(promise: Promise) {
        try {
            if (!isScanning) {
                promise.resolve(Arguments.createMap().apply {
                    putBoolean("scanning", false)
                })
                return
            }

            bluetoothLeScanner?.stopScan(scanCallback)
            isScanning = false

            Log.d(TAG, "BLE beacon scanning stopped")
            NativeLogger.info(TAG, "BLE beacon scanning stopped")

            sendEvent("onScanningStateChange", Arguments.createMap().apply {
                putBoolean("scanning", false)
            })

            promise.resolve(Arguments.createMap().apply {
                putBoolean("scanning", false)
            })

        } catch (e: Exception) {
            promise.reject("STOP_ERROR", "Failed to stop scanning: ${e.message}")
        }
    }

    @ReactMethod
    fun isCurrentlyScanning(promise: Promise) {
        promise.resolve(Arguments.createMap().apply {
            putBoolean("scanning", isScanning)
        })
    }

    @ReactMethod
    fun getBluetoothState(promise: Promise) {
        try {
            val state = when {
                bluetoothAdapter == null -> "unsupported"
                !bluetoothAdapter!!.isEnabled -> "poweredOff"
                else -> "poweredOn"
            }

            val result = Arguments.createMap().apply {
                putString("state", state)
                putBoolean("scanning", isScanning)
            }

            promise.resolve(result)
        } catch (e: Exception) {
            promise.reject("STATE_ERROR", "Failed to get Bluetooth state: ${e.message}")
        }
    }

    // Required for New Architecture event emitter
    @ReactMethod
    fun addListener(eventName: String) {
        // Keep: Required for RN built-in Event Emitter Calls
    }

    @ReactMethod
    fun removeListeners(count: Int) {
        // Keep: Required for RN built-in Event Emitter Calls
    }

    private fun processScanResult(result: ScanResult) {
        try {
            val scanRecord = result.scanRecord ?: return

            // Log ALL discovered BLE devices to native logcat only (verbose)
            val deviceInfo = buildString {
                append("BLE Device Discovered:\n")
                append("  Name: ${result.device.name ?: "No Name"}\n")
                append("  Address: ${result.device.address}\n")
                append("  RSSI: ${result.rssi}")
            }
            Log.d(TAG, deviceInfo)

            val manufacturerData = scanRecord.manufacturerSpecificData

            // Check if manufacturer data exists
            if (manufacturerData == null || manufacturerData.size() == 0) {
                Log.d(TAG, "  No manufacturer data")
                return
            }

            // Log manufacturer data
            val mfgInfo = buildString {
                append("  Manufacturer Data:")
                for (i in 0 until manufacturerData.size()) {
                    val companyId = manufacturerData.keyAt(i)
                    val data = manufacturerData.valueAt(i)
                    val hexData = data.joinToString("") { String.format("%02X", it) }
                    append("\n    Company ID: 0x${String.format("%04X", companyId)} Data: $hexData")
                }
            }
            Log.d(TAG, mfgInfo)

            // Look for Phoenix beacon in Samsung (Android) or Apple (iOS) manufacturer data
            val phoenixData = manufacturerData.get(PHOENIX_COMPANY_ID_SAMSUNG)
                ?: manufacturerData.get(PHOENIX_COMPANY_ID_APPLE)
                ?: return

            // Validate Phoenix beacon format: magic (2 bytes) + data (20 bytes) = 22 bytes
            if (phoenixData.size != 22) {
                Log.d(TAG, "Not Phoenix beacon - size: ${phoenixData.size} bytes (expected 22)")
                return
            }

            // Check for Phoenix magic number (0x5048 = "PH")
            val magic = ((phoenixData[1].toInt() and 0xFF) shl 8) or (phoenixData[0].toInt() and 0xFF)
            if (magic != 0x5048) {
                Log.d(TAG, "Not Phoenix beacon - magic: 0x${String.format("%04X", magic)} (expected 0x5048)")
                return
            }

            val beaconDetected = buildString {
                append("\n========================================\n")
                append("*** PHOENIX BEACON DETECTED ***\n")
                append("========================================\n")
                append("Device: ${result.device.name ?: result.device.address}\n")
                append("RSSI: ${result.rssi} dBm")
            }
            Log.d(TAG, beaconDetected)
            NativeLogger.info(TAG, beaconDetected)

            // Extract the 20-byte beacon data (skip first 2 bytes which are magic)
            val beaconDataBytes = ByteArray(20)
            System.arraycopy(phoenixData, 2, beaconDataBytes, 0, 20)

            // For distance estimation, use a default measured power
            // This can be calibrated later or added to the beacon format
            val measuredPower = -59 // Default RSSI at 1 meter

            // Convert to hex string
            val hexData = beaconDataBytes.joinToString("") { String.format("%02X", it) }

            Log.d(TAG, "  Beacon Data: $hexData")
            NativeLogger.info(TAG, "  Beacon Data: $hexData")

            // Send event to React Native
            val event = Arguments.createMap().apply {
                putString("deviceId", result.device.address)
                putString("deviceName", result.device.name ?: result.device.address)
                putString("beaconData", hexData)
                putInt("rssi", result.rssi)
                putInt("measuredPower", measuredPower)
                putDouble("timestamp", System.currentTimeMillis().toDouble())
            }

            sendEvent("onBeaconDiscovered", event)

        } catch (e: Exception) {
            val errorMsg = "Error processing scan result: ${e.message}"
            Log.e(TAG, errorMsg)
            NativeLogger.error(TAG, errorMsg)
        }
    }

    private fun sendEvent(eventName: String, params: WritableMap) {
        reactApplicationContext
            .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
            .emit(eventName, params)
    }

    private fun getScanErrorMessage(errorCode: Int): String {
        return when (errorCode) {
            ScanCallback.SCAN_FAILED_ALREADY_STARTED -> "Scan already started"
            ScanCallback.SCAN_FAILED_APPLICATION_REGISTRATION_FAILED -> "App registration failed"
            ScanCallback.SCAN_FAILED_INTERNAL_ERROR -> "Internal error"
            ScanCallback.SCAN_FAILED_FEATURE_UNSUPPORTED -> "Feature not supported"
            else -> "Unknown error: $errorCode"
        }
    }
}

File: apps/receiver/native-modules/ios/BLEBeaconScanner.swift
import Foundation
import React
import CoreBluetooth

@objc(BLEBeaconScanner)
class BLEBeaconScanner: RCTEventEmitter {

  private var centralManager: CBCentralManager?
  private var isScanning: Bool = false
  private var hasListeners: Bool = false

  override init() {
    super.init()
  }

  @objc
  static override func requiresMainQueueSetup() -> Bool {
    return true
  }

  override func supportedEvents() -> [String]! {
    return ["onBeaconDiscovered", "onScanningStateChange"]
  }

  override func startObserving() {
    hasListeners = true
  }

  override func stopObserving() {
    hasListeners = false
  }

  // MARK: - Initialization

  @objc
  func initializeScanner(_ resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
    DispatchQueue.main.async {
      if self.centralManager == nil {
        self.centralManager = CBCentralManager(delegate: self, queue: nil)
      }

      // Wait for state to update
      DispatchQueue.main.asyncAfter(deadline: .now() + 0.5) {
        let state = self.centralManager?.state ?? .unknown
        resolve([
          "initialized": true,
          "state": self.stateToString(state)
        ])
      }
    }
  }

  // MARK: - Scanning

  @objc
  func startScanning(_ resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
    DispatchQueue.main.async {
      guard let centralManager = self.centralManager else {
        reject("NOT_INITIALIZED", "Central manager not initialized", nil)
        return
      }

      if centralManager.state != .poweredOn {
        reject("BLUETOOTH_OFF", "Bluetooth is not powered on: \(self.stateToString(centralManager.state))", nil)
        return
      }

      if self.isScanning {
        resolve(["scanning": true])
        return
      }

      // Start scanning for all devices (no service UUID filter)
      // We'll filter in the discovery callback
      centralManager.scanForPeripherals(
        withServices: nil,
        options: [
          CBCentralManagerScanOptionAllowDuplicatesKey: true
        ]
      )

      self.isScanning = true

      print("BLE beacon scanning started")

      if self.hasListeners {
        self.sendEvent(withName: "onScanningStateChange", body: [
          "scanning": true
        ])
      }

      resolve(["scanning": true])
    }
  }

  @objc
  func stopScanning(_ resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
    DispatchQueue.main.async {
      guard let centralManager = self.centralManager else {
        reject("NOT_INITIALIZED", "Central manager not initialized", nil)
        return
      }

      if !self.isScanning {
        resolve(["scanning": false])
        return
      }

      centralManager.stopScan()
      self.isScanning = false

      print("BLE beacon scanning stopped")

      if self.hasListeners {
        self.sendEvent(withName: "onScanningStateChange", body: [
          "scanning": false
        ])
      }

      resolve(["scanning": false])
    }
  }

  @objc
  func isCurrentlyScanning(_ resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
    resolve(["scanning": self.isScanning])
  }

  @objc
  func getBluetoothState(_ resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) {
    guard let centralManager = self.centralManager else {
      resolve(["state": "uninitialized"])
      return
    }

    resolve(["state": self.stateToString(centralManager.state)])
  }

  // MARK: - Helper Methods

  private func stateToString(_ state: CBManagerState) -> String {
    switch state {
    case .unknown:
      return "unknown"
    case .resetting:
      return "resetting"
    case .unsupported:
      return "unsupported"
    case .unauthorized:
      return "unauthorized"
    case .poweredOff:
      return "poweredOff"
    case .poweredOn:
      return "poweredOn"
    @unknown default:
      return "unknown"
    }
  }

  private func parseIBeaconData(_ manufacturerData: Data) -> [String: Any]? {
    // Manufacturer data format: [Company ID (2 bytes)] + [iBeacon data]
    // iBeacon data: [Type (0x02)] + [Length (0x15)] + [UUID (16)] + [Major (2)] + [Minor (2)] + [TxPower (1)]

    guard manufacturerData.count >= 25 else {
      return nil
    }

    // Check Company ID (0x004C for Apple, little-endian)
    let companyID = UInt16(manufacturerData[0]) | (UInt16(manufacturerData[1]) << 8)
    guard companyID == 0x004C else {
      return nil
    }

    // Check iBeacon type and length
    guard manufacturerData[2] == 0x02 && manufacturerData[3] == 0x15 else {
      return nil
    }

    // Extract 20-byte beacon data from iBeacon format
    // UUID (16 bytes) + Major (2 bytes) + Minor (2 bytes) = 20 bytes
    var beaconData = Data()
    beaconData.append(manufacturerData[4..<20])  // UUID (16 bytes)
    beaconData.append(manufacturerData[20..<22]) // Major (2 bytes)
    beaconData.append(manufacturerData[22..<24]) // Minor (2 bytes)

    // Extract measured power (TX power at 1m)
    let measuredPower = Int8(bitPattern: manufacturerData[24])

    // Convert to hex string
    let hexData = beaconData.map { String(format: "%02X", $0) }.joined()

    return [
      "beaconData": hexData,
      "dataLength": beaconData.count,
      "measuredPower": measuredPower
    ]
  }
}

// MARK: - CBCentralManagerDelegate

extension BLEBeaconScanner: CBCentralManagerDelegate {

  func centralManagerDidUpdateState(_ central: CBCentralManager) {
    let state = stateToString(central.state)
    print("BLE Central Manager state changed: \(state)")

    // If scanning and state changed to non-poweredOn, stop scanning
    if central.state != .poweredOn && isScanning {
      isScanning = false
      if hasListeners {
        sendEvent(withName: "onScanningStateChange", body: [
          "scanning": false
        ])
      }
    }
  }

  func centralManager(_ central: CBCentralManager, didDiscover peripheral: CBPeripheral, advertisementData: [String : Any], rssi RSSI: NSNumber) {

    // Extract manufacturer data
    guard let manufacturerData = advertisementData[CBAdvertisementDataManufacturerDataKey] as? Data else {
      return
    }

    // Parse iBeacon data
    guard let beaconInfo = parseIBeaconData(manufacturerData) else {
      return
    }

    print("🔍 PHOENIX BEACON DETECTED!")
    print("  Device: \(peripheral.name ?? peripheral.identifier.uuidString)")
    print("  RSSI: \(RSSI)")
    print("  Beacon Data: \(beaconInfo["beaconData"] ?? "")")

    // Send event to React Native
    if hasListeners {
      sendEvent(withName: "onBeaconDiscovered", body: [
        "deviceId": peripheral.identifier.uuidString,
        "deviceName": peripheral.name ?? "Unknown",
        "beaconData": beaconInfo["beaconData"] ?? "",
        "rssi": RSSI.intValue,
        "measuredPower": beaconInfo["measuredPower"] ?? -59,
        "timestamp": Date().timeIntervalSince1970 * 1000
      ])
    }
  }
}
